src/dialects/mssql/index.js

```diff
@@ -1,207 +1,399 @@
+    return new QueryCompiler(this, ...arguments);
   },
 
   schemaCompiler() {
-    return new SchemaCompiler(this, ...arguments)
+    return new SchemaCompiler(this, ...arguments);
   },
 
   tableCompiler() {
-    return new TableCompiler(this, ...arguments)
+    return new TableCompiler(this, ...arguments);
   },
 
   columnCompiler() {
-    return new ColumnCompiler(this, ...arguments)
+    return new ColumnCompiler(this, ...arguments);
   },
 
   wrapIdentifierImpl(value) {
-    return (value !== '*' ? `[${value.replace(/\[/g, '[')}]` : '*')
+    return value !== "*" ? `[${value.replace(/\[/g, "[")}]` : "*";
   },
 
   // Get a raw connection, called by the `pool` whenever a new
   // connection needs to be added to the pool.
   acquireRawConnection() {
-    return new Promise((resolver, rejecter) => {
-      const connection = new this.driver.ConnectionPool(this.connectionSettings);
-      connection.connect((err) => {
+    return new Promise((resolve, reject) => {
+      const config = Object.assign(
+        {},
+        {
+          server: this.config.connection.host,
+          userName: this.config.connection.user,
+          options: {}
+        },
+        this.config.connection
+      );
+
+      Object.assign(
+        config.options,
+        { database: this.config.connection.database },
+        this.config.connection.options
+      );
+
+      const connection = new tedious.Connection(config);
+      debug("connection::connection new connection requested");
+
+      connection.once("connect", err => {
         if (err) {
-          return rejecter(err)
+          debug("connection::connect error: %s", err.message);
+          return reject(err);
         }
-        connection.on('error', (err) => {
-          connection.__knex__disposed = err
-        })
-        resolver(connection);
+
+        debug("connection::connect connected to server");
+
+        connection.connected = true;
+        connection.on("error", e => {
+          debug("connection::error message=%s", e.message);
+          connection.__knex__disposed = e;
+          connection.connected = false;
+        });
+
+        connection.once("end", () => {
+          connection.connected = false;
+          connection.__knex__disposed = "Connection to server was terminated.";
+          debug("connection::end connection ended.");
+        });
+
+        return resolve(connection);
       });
     });
   },
 
   validateConnection(connection) {
-    if (connection.connected === true) {
-      return true
-    }
-
-    return false
+    return connection && connection.connected;
   },
 
   // Used to explicitly close a connection, called internally by the pool
   // when a connection times out or the pool is shutdown.
   destroyRawConnection(connection) {
-    return connection.close()
+    debug("connection::destroy");
+    return new Promise((resolve, reject) => {
+      connection.once("end", () => {
+        resolve();
```

Comment:
You need to be careful to make sure that if the connection has been closed due to a connection error, that this promise still resolves. because the `end` event won't trigger when calling `connection.close()` if it's already closed.

We've had this problem with the `mssql` lib (see https://github.com/tediousjs/node-mssql/blob/v5.0.0/lib/tedious.js#L278-L298)